File: go/internal/core/store_factory.go
package core

// Select the store DRIVER once from the environment — fail-loud (mirrors python store_factory + node store_factory).
// DATABASE_URL=postgres://… selects the Postgres driver (behind //go:build postgres); anything else selects
// SQLite (DATABASE_PATH or in-memory). A DATABASE_URL naming a backend whose driver isn't built FAILS LOUD — never a
// silent fallback to SQLite, which would mask a misconfigured production database.

import (
	"os"
	"strings"
)

// storeDriver is the backend behind the KV facade — the raw (ns, rawKey, rawJSON) interface both the sqlite and
// postgres drivers implement, so KV[K,V] is backend-agnostic.
type storeDriver interface {
	get(ns, key string) (string, bool)
	set(ns, key, raw string)
	del(ns, key string)
	all(ns string) []string
	nextID(name string) int
	do(ns, key string, fn func(rawCur string, exists bool) (string, bool))
}

// driver — the one store backend for this process, opened at package init (parity with the previous `var db`).
var driver = selectDriver()

func selectDriver() storeDriver {
	url := os.Getenv("DATABASE_URL")
	if strings.HasPrefix(url, "postgres://") || strings.HasPrefix(url, "postgresql://") {
		return newPostgresDriver(url) // the real driver under -tags postgres; a fail-loud stub in the default build
	}
	return newSqliteDriver()
}

File: python/app_pkg/core/store_factory.py
"""Select the store DRIVER once from the environment — lazy, memoized, fail-loud (mirrors the storage domain's
get_provider()). DATABASE_URL=postgres://… selects the Postgres driver; anything else selects SQLite (DATABASE_PATH
or in-memory). A DATABASE_URL that names a backend whose driver isn't installed FAILS LOUD — never a silent
fallback to SQLite, which would mask a misconfigured production database."""
import os

from .store_sqlite import SqliteDriver

_driver = None


def get_driver():
    """The one store backend for this process (memoized). Scheme-sniff DATABASE_URL; default = SQLite."""
    global _driver
    if _driver is None:
        url = os.getenv("DATABASE_URL", "")
        if url.startswith("postgres://") or url.startswith("postgresql://"):
            try:
                from .store_postgres import PostgresDriver   # lazy: psycopg is an OPTIONAL dep (zero-dep default)
            except ImportError:
                raise RuntimeError(                          # fail loud — NEVER silently fall back to SQLite
                    "DATABASE_URL names Postgres, but the psycopg driver is not installed. "
                    "Install it: pip install 'psycopg[binary]'") from None
            _driver = PostgresDriver(url)
        else:
            _driver = SqliteDriver(os.getenv("DATABASE_PATH") or ":memory:")
    return _driver
